sympy/functions/combinatorial/factorials.py

```diff
@@ -846,18 +846,24 @@ def _eval(self, n, k):
     def eval(cls, n, k):
         n, k = map(sympify, (n, k))
         d = n - k
+        is_int = (n.is_integer and k.is_integer)
         if d.is_zero or k.is_zero:
             return S.One
         elif d.is_zero is False:
             if (k - 1).is_zero:
                 return n
             elif k.is_negative:
-                return S.Zero
-            elif n.is_integer and n.is_nonnegative and d.is_negative:
+                if (is_int == S.true or is_int is None) and (n.is_nonnegative or n < k):
```

Comment:
Using `n < k` here will raise an exception when the truth value of this relation cannot be determined. Example:
```
n = symbols('n', nonzero=True)
k = symbols('k', negative=True)
binomial(n+k, k)   # TypeError: cannot determine truth value of Relational
```

Reply:
@normalhuman we can probably add a try catch around `if` for such cases to raise a `TypeError`, because `n` can be positive or negative (ambiguous).